bin/network-monitor/src/counter.rs

```diff
@@ -90,45 +89,227 @@ async fn fetch_counter_value(
+        "code commitment mismatch: rebuilt={:?}, expected={:?}",
+        account.code().commitment(),
+        expected_code_commitment
+    );
+    anyhow::ensure!(
+        account.vault().root() == expected_vault_root,
+        "vault root mismatch: rebuilt={:?}, expected={:?}",
+        account.vault().root(),
+        expected_vault_root
+    );
+    anyhow::ensure!(
+        account.storage().to_commitment() == expected_storage_commitment,
+        "storage commitment mismatch: rebuilt={:?}, expected={:?}",
+        account.storage().to_commitment(),
+        expected_storage_commitment
+    );
+
+    info!(account.id = %account_id, "fetched wallet account from RPC");
     Ok(Some(account))
 }
 
+/// Build account storage from the storage details returned by the server.
+fn build_account_storage(
+    account_id: AccountId,
+    storage_details: miden_node_proto::generated::rpc::AccountStorageDetails,
+) -> Result<miden_objects::account::AccountStorage> {
+    use miden_objects::account::{AccountStorage, StorageSlot};
+
+    let storage_header = storage_details.header.context("missing storage header")?;
+
+    // Build a map of slot_name -> map entries from the response
+    let map_entries: std::collections::HashMap<String, Vec<(Word, Word)>> = storage_details
+        .map_details
+        .into_iter()
+        .filter_map(|map_detail| {
+            let entries = map_detail.entries?.entries;
+            let converted: Vec<(Word, Word)> = entries
+                .into_iter()
+                .filter_map(|e| {
+                    let key: Word = e.key?.try_into().ok()?;
+                    let value: Word = e.value?.try_into().ok()?;
+                    Some((key, value))
+                })
+                .collect();
+            Some((map_detail.slot_name, converted))
+        })
+        .collect();
+
+    let mut slots = Vec::new();
+    for slot in storage_header.slots {
+        let slot_name = miden_objects::account::StorageSlotName::new(slot.slot_name.clone())
+            .context("invalid slot name")?;
+        let commitment: Word = slot
+            .commitment
+            .context("missing slot commitment")?
+            .try_into()
+            .context("invalid slot commitment")?;
+
+        // slot_type: 0 = Value, 1 = Map
+        let storage_slot = if slot.slot_type == 0 {
+            StorageSlot::with_value(slot_name, commitment)
+        } else {
+            let entries = map_entries.get(&slot.slot_name).cloned().unwrap_or_default();
+            if entries.is_empty() {
+                warn!(
+                    account.id = %account_id,
+                    slot_name = %slot.slot_name,
+                    "storage map slot has no entries (not requested or empty)"
+                );
+            }
+            let storage_map = miden_objects::account::StorageMap::with_entries(entries)
+                .context("failed to create storage map")?;
+            StorageSlot::with_map(slot_name, storage_map)
```

Comment:
Similar to the above comment: I think we can probably just error out if there are any storage maps in the account.